doc/source/reference/c-api/array.rst

```diff
@@ -2884,11 +3172,279 @@ Data-type descriptors
+   common-dtype implementation is associative and commutative!  (Mainly,
+   unsigned and signed integers are not.)
+
+   For guaranteed consistent results DTypes must implement common-Dtype
+   "transitively".  If A promotes B and B promotes C, than A must generally
+   also promote C; where "promotes" means implements the promotion.  (There
+   are some exceptions for abstract DTypes)
+
+   In general this approach always works as long as the most generic dtype
+   is either strictly larger, or compatible with all other dtypes.
+   For example promoting ``float16`` with any other float, integer, or unsigned
+   integer again gives a floating point number.
+
+.. c:function:: PyArray_Descr *PyArray_GetDefaultDescr(PyArray_DTypeMeta *DType)
+
+   Given a DType class, returns the default instance (descriptor).  This checks
+   for a ``singleton`` first and only calls the ``default_descr`` function if
+   necessary.
+
+.. dtype-api_
+   
+Custom Data Types
+-----------------
+
+..versionadded:: 2.0
+
+These functions allow defining custom flexible data types outside of NumPy.  See
+:ref:`NEP 42 <NEP42>` for more details about the rationale and design of the new
+DType system. See the `numpy-user-dtypes repository
+<https://github.com/numpy/numpy-user-dtypes>`_ for a number of example DTypes.
+Also see :ref:`dtypemeta` for documentation on ``PyArray_DTypeMeta`` and
+``PyArrayDTypeMeta_Spec``.
+
+.. c:function:: int PyArrayInitDTypeMeta_FromSpec( \
+                PyArray_DTypeMeta *Dtype, PyArrayDTypeMeta_Spec *spec)
+
+ Initialize a new DType.  It must currently be a static Python C type that is
+ declared as :c:type:`PyArray_DTypeMeta` and not c:type:`PyTypeObject`.
+ Further, it must subclass `np.dtype` and set its type to
+ :c:type:`PyArrayDTypeMeta_Type` (before calling :c:func:`PyType_Ready()`),
+ which has additional fields compared to a normal :c:type:`PyTypeObject`. See
+ the examples in the ``numpy-user-dtypes`` repository for usage with both
+ parametric and non-parametric data types.
+
+.. _dtype-flags:
+
+Flags
+~~~~~
+
+Flags that can be set on the ``PyArrayDTypeMeta_Spec`` to initialize the DType.
+
+.. c:macro:: NPY_DT_ABSTRACT
+
+   Indicates the DType is an abstract "base" DType in a DType hierarchy and
+   should not be directly instantiated.
+
+.. c:macro:: NPY_DT_PARAMETRIC
+
+   Indicates the DType is parametric and does not have a unique singleton
+   instance.
+
+.. c:macro:: NPY_DT_NUMERIC
+
+   Indicates the DType is represents a numerical value.
+
+
+.. _dtype-slots:
+
+Slot IDs and API Function Typedefs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These IDs correspond to slots in the DType API and are used to identify
+implementations of each slot from the items of the ``slots`` array
+member of ``PyArrayDTypeMeta_Spec`` struct.
+
+.. c:macro:: NPY_DT_discover_descr_from_pyobject
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_DiscoverDescrFromPyobject)( \
```

Comment:
This combination of flags and function typedef is rather nice. Use it for the array methods too?

Reply:
Rearranged the arraymethod docs.

Reply:
Thanks, that helps the arraymethod part!